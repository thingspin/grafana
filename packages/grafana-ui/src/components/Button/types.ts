import { AnchorHTMLAttributes, ButtonHTMLAttributes, ComponentType } from 'react';
import { GrafanaTheme, Themeable } from '../../types';

export type ButtonVariant = 'primary' | 'secondary' | 'danger' | 'inverse' | 'transparent' | 'destructive';

export type ButtonSize = 'xs' | 'sm' | 'md' | 'lg';

export interface StyleDeps {
  theme: GrafanaTheme;
  size: ButtonSize;
  variant: ButtonVariant;
  withIcon: boolean;
}

export interface ButtonStyles {
  button: string;
  iconWrap: string;
  icon: string;
}

export interface CommonButtonProps {
  size?: ButtonSize;
  variant?: ButtonVariant;
  /**
   * icon prop is a temporary solution. It accepts legacy icon class names for the icon to be rendered.
   * TODO: migrate to a component when we are going to migrate icons to @grafana/ui
   */
  icon?: string;
  className?: string;
}

export interface LinkButtonProps extends CommonButtonProps, AnchorHTMLAttributes<HTMLAnchorElement> {
  disabled?: boolean;
}
export interface ButtonProps extends CommonButtonProps, ButtonHTMLAttributes<HTMLButtonElement> {}

export interface AbstractButtonProps extends CommonButtonProps, Themeable {
  renderAs: ComponentType<CommonButtonProps> | string;
}
