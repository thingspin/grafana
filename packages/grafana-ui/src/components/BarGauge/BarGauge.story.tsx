import { storiesOf } from '@storybook/react';
import { number, text, boolean } from '@storybook/addon-knobs';
import { BarGauge } from './BarGauge';
import { VizOrientation } from '../../types';
import { withHorizontallyCenteredStory } from '../../utils/storybook/withCenteredStory';
import { renderComponentWithTheme } from '../../utils/storybook/withTheme';

const getKnobs = () => {
  return {
    value: number('value', 70),
    minValue: number('minValue', 0),
    maxValue: number('maxValue', 100),
    threshold1Value: number('threshold1Value', 40),
    threshold1Color: text('threshold1Color', 'orange'),
    threshold2Value: number('threshold2Value', 60),
    threshold2Color: text('threshold2Color', 'red'),
    unit: text('unit', 'ms'),
    decimals: number('decimals', 1),
    horizontal: boolean('horizontal', false),
    lcd: boolean('lcd', false),
  };
};

const BarGaugeStories = storiesOf('UI/BarGauge/BarGauge', module);

BarGaugeStories.addDecorator(withHorizontallyCenteredStory);

BarGaugeStories.add('Simple with basic thresholds', () => {
  const {
    value,
    minValue,
    maxValue,
    threshold1Color,
    threshold2Color,
    threshold1Value,
    threshold2Value,
    unit,
    decimals,
    horizontal,
    lcd,
  } = getKnobs();

  return renderComponentWithTheme(BarGauge, {
    width: 300,
    height: 300,
    value: { text: value.toString(), numeric: value },
    minValue: minValue,
    maxValue: maxValue,
    unit: unit,
    prefix: '',
    postfix: '',
    decimals: decimals,
    orientation: horizontal ? VizOrientation.Horizontal : VizOrientation.Vertical,
    displayMode: lcd ? 'lcd' : 'simple',
    thresholds: [
      { index: 0, value: -Infinity, color: 'green' },
      { index: 1, value: threshold1Value, color: threshold1Color },
      { index: 1, value: threshold2Value, color: threshold2Color },
    ],
  });
});
