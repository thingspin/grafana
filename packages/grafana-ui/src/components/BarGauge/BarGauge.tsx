// Library
import React, { PureComponent, CSSProperties, ReactNode } from 'react';
import tinycolor from 'tinycolor2';

// Utils
import { getColorFromHexRgbOrName, getThresholdForValue, DisplayValue } from '../../utils';

// Types
import { Themeable, TimeSeriesValue, Threshold, VizOrientation } from '../../types';

const BAR_SIZE_RATIO = 0.8;

export interface Props extends Themeable {
  height: number;
  width: number;
  thresholds: Threshold[];
  value: DisplayValue;
  maxValue: number;
  minValue: number;
  orientation: VizOrientation;
  displayMode: 'basic' | 'lcd' | 'gradient';
}

export class BarGauge extends PureComponent<Props> {
  static defaultProps: Partial<Props> = {
    maxValue: 100,
    minValue: 0,
    value: {
      text: '100',
      numeric: 100,
    },
    displayMode: 'lcd',
    orientation: VizOrientation.Horizontal,
    thresholds: [],
  };

  render() {
    switch (this.props.displayMode) {
      case 'lcd':
        return this.renderRetroBars();
      case 'basic':
      case 'gradient':
      default:
        return this.renderBasicAndGradientBars();
    }
  }

  getValueColors(): BarColors {
    const { thresholds, theme, value } = this.props;

    const activeThreshold = getThresholdForValue(thresholds, value.numeric);

    if (activeThreshold !== null) {
      const color = getColorFromHexRgbOrName(activeThreshold.color, theme.type);

      return {
        value: color,
        border: color,
        background: tinycolor(color)
          .setAlpha(0.15)
          .toRgbString(),
      };
    }

    return {
      value: getColorFromHexRgbOrName('gray', theme.type),
      background: getColorFromHexRgbOrName('gray', theme.type),
      border: getColorFromHexRgbOrName('gray', theme.type),
    };
  }

  getValueStyles(value: string, color: string, width: number): CSSProperties {
    const guess = width / (value.length * 1.1);
    const fontSize = Math.min(Math.max(guess, 14), 40);

    return {
      color: color,
      fontSize: fontSize + 'px',
    };
  }

  /*
   * Return width or height depending on viz orientation
   * */
  get size() {
    const { height, width } = this.props;
    return this.isVertical ? height : width;
  }

  get isVertical() {
    return this.props.orientation === VizOrientation.Vertical;
  }

  getBarGradient(maxSize: number): string {
    const { minValue, maxValue, thresholds, value } = this.props;
    const cssDirection = this.isVertical ? '0deg' : '90deg';

    let gradient = '';
    let lastpos = 0;

    for (let i = 0; i < thresholds.length; i++) {
      const threshold = thresholds[i];
      const color = getColorFromHexRgbOrName(threshold.color);
      const valuePercent = Math.min(threshold.value / (maxValue - minValue), 1);
      const pos = valuePercent * maxSize;
      const offset = Math.round(pos - (pos - lastpos) / 2);

      if (gradient === '') {
        gradient = `linear-gradient(${cssDirection}, ${color}, ${color}`;
      } else if (value.numeric < threshold.value) {
        break;
      } else {
        lastpos = pos;
        gradient += ` ${offset}px, ${color}`;
      }
    }

    return gradient + ')';
  }

  renderBasicAndGradientBars(): ReactNode {
    const { height, width, displayMode, maxValue, minValue, value } = this.props;

    const valuePercent = Math.min(value.numeric / (maxValue - minValue), 1);
    const maxSize = this.size * BAR_SIZE_RATIO;
    const barSize = Math.max(valuePercent * maxSize, 0);
    const colors = this.getValueColors();
    const spaceForText = this.isVertical ? width : Math.min(this.size - maxSize, height);
    const valueStyles = this.getValueStyles(value.text, colors.value, spaceForText);
    const isBasic = displayMode === 'basic';

    const containerStyles: CSSProperties = {
      width: `${width}px`,
      height: `${height}px`,
      display: 'flex',
    };

    const barStyles: CSSProperties = {
      borderRadius: '3px',
    };

    if (this.isVertical) {
      // Custom styles for vertical orientation
      containerStyles.flexDirection = 'column';
      containerStyles.justifyContent = 'flex-end';
      barStyles.transition = 'height 1s';
      barStyles.height = `${barSize}px`;
      barStyles.width = `${width}px`;
      if (isBasic) {
        // Basic styles
        barStyles.background = `${colors.background}`;
        barStyles.border = `1px solid ${colors.border}`;
        barStyles.boxShadow = `0 0 4px ${colors.border}`;
      } else {
        // Gradient styles
        barStyles.background = this.getBarGradient(maxSize);
      }
    } else {
      // Custom styles for horizontal orientation
      containerStyles.flexDirection = 'row-reverse';
      containerStyles.justifyContent = 'flex-end';
      containerStyles.alignItems = 'center';
      barStyles.transition = 'width 1s';
      barStyles.height = `${height}px`;
      barStyles.width = `${barSize}px`;
      barStyles.marginRight = '10px';

      if (isBasic) {
        // Basic styles
        barStyles.background = `${colors.background}`;
        barStyles.border = `1px solid ${colors.border}`;
        barStyles.boxShadow = `0 0 4px ${colors.border}`;
      } else {
        // Gradient styles
        barStyles.background = this.getBarGradient(maxSize);
      }
    }

    return (
      <div style={containerStyles}>
        <div className="bar-gauge__value" style={valueStyles}>
          {value.text}
        </div>
        <div style={barStyles} />
      </div>
    );
  }

  getCellColor(positionValue: TimeSeriesValue): CellColors {
    const { thresholds, theme, value } = this.props;
    const activeThreshold = getThresholdForValue(thresholds, positionValue);

    if (activeThreshold !== null) {
      const color = getColorFromHexRgbOrName(activeThreshold.color, theme.type);

      // if we are past real value the cell is not "on"
      if (value === null || (positionValue !== null && positionValue > value.numeric)) {
        return {
          background: tinycolor(color)
            .setAlpha(0.15)
            .toRgbString(),
          border: 'transparent',
          isLit: false,
        };
      } else {
        return {
          background: tinycolor(color)
            .setAlpha(0.85)
            .toRgbString(),
          backgroundShade: tinycolor(color)
            .setAlpha(0.55)
            .toRgbString(),
          border: tinycolor(color)
            .setAlpha(0.9)
            .toRgbString(),
          isLit: true,
        };
      }
    }

    return {
      background: 'gray',
      border: 'gray',
    };
  }

  renderRetroBars(): ReactNode {
    const { height, width, maxValue, minValue, value } = this.props;

    const valueRange = maxValue - minValue;
    const maxSize = this.size * BAR_SIZE_RATIO;
    const cellSpacing = 5;
    const cellCount = maxSize / 20;
    const cellSize = (maxSize - cellSpacing * cellCount) / cellCount;
    const colors = this.getValueColors();
    const spaceForText = this.isVertical ? width : Math.min(this.size - maxSize, height);
    const valueStyles = this.getValueStyles(value.text, colors.value, spaceForText);

    const containerStyles: CSSProperties = {
      width: `${width}px`,
      height: `${height}px`,
      display: 'flex',
    };

    if (this.isVertical) {
      containerStyles.flexDirection = 'column-reverse';
      containerStyles.alignItems = 'center';
      valueStyles.marginBottom = '20px';
    } else {
      containerStyles.flexDirection = 'row';
      containerStyles.alignItems = 'center';
      valueStyles.marginLeft = '20px';
    }

    const cells: JSX.Element[] = [];

    for (let i = 0; i < cellCount; i++) {
      const currentValue = (valueRange / cellCount) * i;
      const cellColor = this.getCellColor(currentValue);
      const cellStyles: CSSProperties = {
        borderRadius: '2px',
      };

      if (cellColor.isLit) {
        cellStyles.boxShadow = `0 0 4px ${cellColor.border}`;
        cellStyles.backgroundImage = `radial-gradient(${cellColor.background} 10%, ${cellColor.backgroundShade})`;
      } else {
        cellStyles.backgroundColor = cellColor.background;
      }

      if (this.isVertical) {
        cellStyles.height = `${cellSize}px`;
        cellStyles.width = `${width}px`;
        cellStyles.marginTop = `${cellSpacing}px`;
      } else {
        cellStyles.width = `${cellSize}px`;
        cellStyles.height = `${height}px`;
        cellStyles.marginRight = `${cellSpacing}px`;
      }

      cells.push(<div style={cellStyles} />);
    }

    return (
      <div style={containerStyles}>
        {cells}
        <div className="bar-gauge__value" style={valueStyles}>
          {value.text}
        </div>
      </div>
    );
  }
}

interface BarColors {
  value: string;
  background: string;
  border: string;
}

interface CellColors {
  background: string;
  backgroundShade?: string;
  border: string;
  isLit?: boolean;
}
