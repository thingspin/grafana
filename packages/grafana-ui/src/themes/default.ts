import { GrafanaThemeCommons } from '../types/theme';

const theme: GrafanaThemeCommons = {
  name: 'Grafana Default',
  typography: {
    fontFamily: {
      sansSerif: "'Roboto', Helvetica, Arial, sans-serif",
      monospace: "Menlo, Monaco, Consolas, 'Courier New', monospace",
    },
    size: {
      root: '14px',
      base: '13px',
      xs: '10px',
      s: '12px',
      m: '14px',
      l: '18px',
    },
    heading: {
      h1: '28px',
      h2: '24px',
      h3: '21px',
      h4: '18px',
      h5: '16px',
      h6: '14px',
    },
    weight: {
      light: 300,
      normal: 400,
      semibold: 500,
    },
    lineHeight: {
      xs: 1,
      s: 1.1,
      m: 4 / 3,
      l: 1.5,
    },
  },
  breakpoints: {
    xs: '0',
    s: '544px',
    m: '768px',
    l: '992px',
    xl: '1200px',
  },
  spacing: {
    xs: '0',
    s: '3px',
    m: '14px',
    l: '21px',
    gutter: '30px',
  },
  border: {
    radius: {
      xs: '2px',
      s: '3px',
      m: '5px',
    },
    width: {
      s: '1px',
    },
  },
};

export default theme;
