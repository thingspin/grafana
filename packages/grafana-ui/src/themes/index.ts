import { ThemeContext, withTheme } from './ThemeContext';
import { getTheme, mockTheme } from './getTheme';

export { ThemeContext, withTheme, mockTheme, getTheme };
