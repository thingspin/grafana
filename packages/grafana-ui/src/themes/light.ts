import tinycolor from 'tinycolor2';
import defaultTheme from './default';
import { GrafanaTheme, GrafanaThemeType } from '../types/theme';

const basicColors = {
  black: '#000000',
  white: '#ffffff',
  dark1: '#1e2028',
  dark2: '#41444b',
  dark3: '#303133', // not used in light theme
  dark4: '#35373f', // not used in light theme
  dark5: '#41444b', // not used in light theme
  dark6: '#41444b', // not used in light theme
  dark7: '#41444b', // not used in light theme
  dark8: '#2f2f32', // not used in light theme
  dark9: '#343436', // not used in light theme
  dark10: '#424345', // not used in light theme
  gray1: '#52545c',
  gray2: '#767980',
  gray3: '#acb6bf',
  gray4: '#c7d0d9',
  gray5: '#dde4ed',
  gray6: '#e9edf2',
  gray7: '#f7f8fa',
  grayBlue: '#212327', // not used in light theme
  blueBase: '#3274d9',
  blueShade: '#1f60c4',
  blueLight: '#5794f2',
  blueFaint: '#f5f9ff',
  redBase: '#e02f44',
  redShade: '#c4162a',
  greenBase: '#3eb15b',
  greenShade: '#369b4f',
  blue: '#0083b3',
  red: '#d44939',
  yellow: '#ff851b',
  purple: '#9954bb',
  variable: '#007580',
  orange: '#ff7941',
};

const lightTheme: GrafanaTheme = {
  ...defaultTheme,
  type: GrafanaThemeType.Light,
  name: 'Grafana Light',
  colors: {
    ...basicColors,
    variable: basicColors.blue,
    inputBlack: '#09090b',
    queryRed: basicColors.red,
    queryGreen: basicColors.greenBase,
    queryPurple: basicColors.purple,
    queryKeyword: basicColors.blue,
    queryOrange: basicColors.orange,
    online: basicColors.greenShade,
    warn: '#f79520',
    critical: basicColors.redShade,
    bodyBg: basicColors.gray7,
    pageBg: basicColors.gray7,
    bodyColor: basicColors.gray1,
    textColor: basicColors.gray1,
    textColorStrong: basicColors.dark2,
    textColorWeak: basicColors.gray2,
    textColorEmphasis: basicColors.gray5,
    textColorFaint: basicColors.dark4,
    linkColor: basicColors.gray1,
    linkColorDisabled: new tinycolor(basicColors.gray1).lighten(30).toString(),
    linkColorHover: new tinycolor(basicColors.gray1).darken(20).toString(),
    linkColorExternal: basicColors.blueLight,
    headingColor: basicColors.gray1,
  },
  background: {
    dropdown: basicColors.white,
    scrollbar: basicColors.gray5,
    scrollbar2: basicColors.gray5,
  },
};

export default lightTheme;
