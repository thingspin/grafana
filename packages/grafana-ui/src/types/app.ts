import { ComponentClass } from 'react';
import { NavModel } from './navModel';
import { PluginMeta, PluginIncludeType, GrafanaPlugin, KeyValue } from './plugin';

export interface AppRootProps<T = KeyValue> {
  meta: AppPluginMeta<T>;

  path: string; // The URL path to this page
  query: KeyValue; // The URL query parameters

  /**
   * Pass the nav model to the container... is there a better way?
   */
  onNavChanged: (nav: NavModel) => void;
}

export interface AppPluginMeta<T = KeyValue> extends PluginMeta<T> {
  // TODO anything specific to apps?
}

export class AppPlugin<T = KeyValue> extends GrafanaPlugin<AppPluginMeta<T>> {
  // Content under: /a/${plugin-id}/*
  root?: ComponentClass<AppRootProps<T>>;
  rootNav?: NavModel; // Initial navigation model

  // Old style pages
  angularPages?: { [component: string]: any };

  /**
   * Called after the module has loaded, and before the app is used.
   * This function may be called multiple times on the same instance.
   * The first time, `this.meta` will be undefined
   */
  init(meta: AppPluginMeta) {}

  /**
   * Set the component displayed under:
   *   /a/${plugin-id}/*
   */
  setRootPage(root: ComponentClass<AppRootProps<T>>, rootNav?: NavModel) {
    this.root = root;
    this.rootNav = rootNav;
    return this;
  }

  setComponentsFromLegacyExports(pluginExports: any) {
    if (pluginExports.ConfigCtrl) {
      this.angularConfigCtrl = pluginExports.ConfigCtrl;
    }

    const { meta } = this;
    if (meta && meta.includes) {
      for (const include of meta.includes) {
        const { type, component } = include;
        if (type === PluginIncludeType.page && component) {
          const exp = pluginExports[component];
          if (!exp) {
            console.warn('App Page uses unknown component: ', component, meta);
            continue;
          }
          if (!this.angularPages) {
            this.angularPages = {};
          }
          this.angularPages[component] = exp;
        }
      }
    }
  }
}
