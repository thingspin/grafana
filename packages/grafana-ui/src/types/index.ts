export * from './data';
export * from './time';
export * from './panel';
export * from './plugin';
export * from './datasource';
export * from './theme';
export * from './threshold';
