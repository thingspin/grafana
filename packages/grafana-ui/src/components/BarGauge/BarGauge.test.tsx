import React from 'react';
import { shallow } from 'enzyme';
import { BarGauge, Props } from './BarGauge';
import { VizOrientation } from '../../types';
import { getTheme } from '../../themes';

jest.mock('jquery', () => ({
  plot: jest.fn(),
}));

const setup = (propOverrides?: object) => {
  const props: Props = {
    maxValue: 100,
    minValue: 0,
    displayMode: 'basic',
    thresholds: [{ index: 0, value: -Infinity, color: '#7EB26D' }],
    height: 300,
    width: 300,
    value: {
      text: '25',
      numeric: 25,
    },
    theme: getTheme(),
    orientation: VizOrientation.Horizontal,
  };

  Object.assign(props, propOverrides);

  const wrapper = shallow(<BarGauge {...props} />);
  const instance = wrapper.instance() as BarGauge;

  return {
    instance,
    wrapper,
  };
};

describe('Get font color', () => {
  it('should get first threshold color when only one threshold', () => {
    const { instance } = setup({ thresholds: [{ index: 0, value: -Infinity, color: '#7EB26D' }] });

    expect(instance.getValueColors().value).toEqual('#7EB26D');
  });

  it('should get the threshold color if value is same as a threshold', () => {
    const { instance } = setup({
      thresholds: [
        { index: 2, value: 75, color: '#6ED0E0' },
        { index: 1, value: 10, color: '#EAB839' },
        { index: 0, value: -Infinity, color: '#7EB26D' },
      ],
    });

    expect(instance.getValueColors().value).toEqual('#EAB839');
  });
});

describe('Render BarGauge with basic options', () => {
  it('should render', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
