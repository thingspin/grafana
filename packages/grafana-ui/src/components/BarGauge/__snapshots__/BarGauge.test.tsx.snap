// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render BarGauge with basic options should render 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "row-reverse",
      "height": "300px",
      "justifyContent": "flex-end",
      "width": "300px",
    }
  }
>
  <div
    className="bar-gauge__value"
    style={
      Object {
        "color": "#7EB26D",
        "fontSize": "27.27272727272727px",
      }
    }
  >
    25
  </div>
  <div
    style={
      Object {
        "background": "rgba(126, 178, 109, 0.15)",
        "border": "1px solid #7EB26D",
        "borderRadius": "3px",
        "boxShadow": "0 0 4px #7EB26D",
        "height": "300px",
        "marginRight": "10px",
        "transition": "width 1s",
        "width": "60px",
      }
    }
  />
</div>
`;
